### PR TITLE
Fix for (slow) memory leak in ClientState

### PR DESCRIPTION
**Pull request by raboof**
`ClientState` registered a `whenTerminated` callback to close the client when
the actor system terminates. However, this `whenTerminated` callback is never
removed, meaning it will hold on to the `ClientState`, causing a (small)
memory leak each time a client is instantiated.

This PR is intended to fix the problem, but doesn't work due to what
looks like an extension initialization deadlock. That reminds of
https://github.com/akka/akka/pull/28046 but it also seems present when I
switch to master, which is on Akka 2.6

Refs https://github.com/akka/akka-grpc/issues/1257

### CODE REVIEW

File: runtime/src/main/scala/akka/grpc/scaladsl/Grpc.scala [hunk 1]
```diff
@@ -0,0 +1,41 @@
...
+import java.util.concurrent.ConcurrentHashMap
+
+/** INTERNAL API */
+@InternalApi
+class GrpcImpl(system: ExtendedActorSystem) extends Extension {
+  private val clients = new ConcurrentHashMap[ClientState, Unit]
+
+  CoordinatedShutdown(system).addTask("before-actor-system-terminate", "close-grpc-clients") { () =>
+    implicit val ec = system.dispatcher
+    Future.sequence(clients.keySet().asScala.map(_.close())).map(_ => Done)
```
Comment (johanandren) [inline runtime/src/main/scala/akka/grpc/scaladsl/Grpc.scala:20]: I think with sequence if one future op fails it won't execute the next so maybe wrap the close with a try just for good measure?

File: runtime/src/main/scala/akka/grpc/scaladsl/Grpc.scala [hunk 2]
```diff
@@ -0,0 +1,41 @@
...
+import akka.Done
+import akka.actor.{ CoordinatedShutdown, ExtendedActorSystem, Extension, ExtensionId, ExtensionIdProvider }
+import akka.annotation.InternalApi
+import akka.grpc.internal.ClientState
+
+import java.util.concurrent.ConcurrentHashMap
+
+/** INTERNAL API */
+@InternalApi
+class GrpcImpl(system: ExtendedActorSystem) extends Extension {
```
Comment (johanandren) [inline runtime/src/main/scala/akka/grpc/scaladsl/Grpc.scala:15]: If you liked it, make it final

File: runtime/src/main/scala/akka/grpc/scaladsl/Grpc.scala [hunk 3]
```diff
@@ -0,0 +1,41 @@
...
+
+  /** INTERNAL API */
+  @InternalApi
+  def deregisterClient(client: ClientState): Unit =
+    clients.remove(client)
+}
+
+/** INTERNAL API */
+@InternalApi
+object Grpc extends ExtensionId[GrpcImpl] with ExtensionIdProvider {
```
Comment (johanandren) [inline runtime/src/main/scala/akka/grpc/scaladsl/Grpc.scala:36]: can be `private[grpc]` perhaps?

**Review by johanandren — APPROVED** [review]
Lookin good, left some nitpicking as a weekend gift.

**Review by johanandren — APPROVED** [review]

**Review by ignasi35 — APPROVED** [review]
LGTM

**Comment by ignasi35** [comment]
Hi @raboof, the branch for this PR starts on the commit tagged `1.0.2`. Did you want to release a `1.0.3` with this fix?

**Comment by raboof** [comment]
> Hi @raboof, the branch for this PR starts on the commit tagged `1.0.2`. Did you want to release a `1.0.3` with this fix?

No, I just started there because that was the version we encountered the leak with. This way we *could* merge it into a 1.0.x branch, but I don't see a strong reason to have such a branch.

**Comment by ignasi35** [comment]
workflow issue is unrelated to this PR (See #1259 )

**Pull request merged**